spacy/util.py

```diff
@@ -256,20 +258,85 @@ def lang_class_is_loaded(lang: str) -> bool:
     return lang in registry.languages
 
 
+def find_matching_language(lang: str) -> Optional[str]:
+    """
+    Given an IETF language code, find a supported spaCy language that is a
+    close match for it (according to Unicode CLDR language-matching rules).
+    This allows for language aliases, ISO 639-2 codes, more detailed language
+    tags, and close matches.
+
+    Returns the language code if a matching language is available, or None
+    if there is no matching language.
+
+    >>> find_matching_language('en')
+    'en'
+    >>> find_matching_language('pt-BR')  # Brazilian Portuguese
+    'pt'
+    >>> find_matching_language('fra')  # an ISO 639-2 code for French
+    'fr'
+    >>> find_matching_language('iw')  # obsolete alias for Hebrew
+    'he'
+    >>> find_matching_language('no')  # Norwegian
+    'nb'
+    >>> find_matching_language('mo')  # old code for ro-MD
+    'ro'
+    >>> find_matching_language('zh-Hans')  # Simplified Chinese
+    'zh'
+    >>> find_matching_language('zxx')
+    None
+    """
+    import spacy.lang  # noqa: F401
+    if lang == 'xx':
+        return 'xx'
+
+    # Find out which language modules we have
+    possible_languages = []
+    for modinfo in pkgutil.iter_modules(spacy.lang.__path__):
+        code = modinfo.name
+        if code == 'xx':
+            # Temporarily make 'xx' into a valid language code
+            code = 'mul'
+        try:
+            language = langcodes.Language.get(code)
+            if language.is_valid():
+                possible_languages.append(code)
+        except langcodes.tag_parser.LanguageTagError:
+            pass
```

Comment:
Is there a way to do this with a conditional, rather than the try/except?

It's not _really_ a problem because we're only catching the specific library error, but I'd still prefer to have conditions expressed with conditionals, and only use try/except for exceptional branches.

Reply:
I admit that this is a wonky way to find out which languages we support. The problem is that if I'm iterating the modules within `spacy.lang`, I'm going to get things like `tokenizer_exceptions`. I can't ask `langcodes.Language.get('tokenizer_exceptions')` if it's valid, because it gets a parse error first.

A way to not encounter this exceptional case would be, perhaps, to put a list of supported languages in `spacy/lang/__init__.py`.

Reply:
...and that would also cope with the possibility where a new submodule of `spacy.lang` just happens to parse as a language tag.

Reply:
We could also make the `try` bit even shorter, which I think further clarifies the flow: 
```suggestion
        language = None
        try:
            language = langcodes.Language.get(code)
        except langcodes.tag_parser.LanguageTagError:
            pass
        if language is not None and language.is_valid():
            possible_languages.append(code)
```
Alternatively, `langcodes` could expose a method that deals with the `try`-`catch` internally, something like `is_parsable`.

Reply:
I do like that this iterates over the actual code base instead of working with a fixed list that we may or may not forget to update at some point ;-)

Reply:
I could make a top-level `langcodes.is_valid_tag()` that takes in a string, and returns false if it doesn't parse _or_ if it parses to an invalid code.

If that's the plan, then we just need to remember not to make submodules that are accidentally valid tags, which would involve the first component being 2-3 letters. I'm imagining a hypothetical module like "no_support.py" where `no_support` parses as a variant of Norwegian.

Reply:
@rspeer : maybe that's a note that you could add to the [dev docs](https://github.com/explosion/spaCy/tree/master/extra/DEVELOPER_DOCS)?